docs/docs/adrs/adr-015-partial-set-security.md

```diff
@@ -0,0 +1,308 @@
+        } else {
+            // remove this validator
+            updates = append(updates, abci.ValidatorUpdate{
+                PubKey: val.PubKey,
+                Power:  0,
+            })
+        }
+    }
+
+    return updates
+}
+```
+And then we change `QueueVSCPackets` to:
+
+```
+// QueueVSCPackets queues latest validator updates for every registered consumer chain
+func (k Keeper) QueueVSCPackets(ctx sdk.Context) {
+    valUpdateID := k.GetValidatorSetUpdateId(ctx) // current valset update ID
+    // Get the validator updates from the staking module.
+    // Note: GetValidatorUpdates panics if the updates provided by the x/staking module
+    // of cosmos-sdk is invalid.
+    valUpdates := k.stakingKeeper.GetValidatorUpdates(ctx)
+
+    for _, chain := range k.GetAllConsumerChains(ctx) {
+        // Apply the key assignment to the validator updates.
+        valUpdates := k.GetPartialSetValidators(ctx, chain.ChainId, valUpdates)
+        valUpdates := k.MustApplyKeyAssignmentToValUpdates(ctx, chain.ChainId, valUpdates)
+        ...
+```
+
+Note that we use `0` to remove validators that are not opted in as described [here](https://docs.cometbft.com/v0.37/spec/abci/abci++_methods#endblock) and as is done for the [assignment of consumer keys](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/provider/keeper/key_assignment.go#L525).
+
+### How do we distribute rewards?
+Currently, rewards are distributed as follows: The consumer [periodically sends rewards](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/consumer/keeper/distribution.go#L148) on the provider `ConsumerRewardsPool` address.
+The provider then [transfers those rewards to the fee collector address](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/provider/keeper/distribution.go#L77) and those transferred rewards are distributed to validators and delegators.
+
+In PSS, we distribute rewards only to validators that actually validate the consumer chain. To do this, we have a pool associated with each consumer chain and consumers IBC transfer the rewards to this pool.
+Then, because we do not use the canonical fee pool and hence rewards are not automatically distributed, we will create a new and modified version of [AllocateTokens](https://github.com/cosmos/cosmos-sdk/blob/v0.47.7/x/distribution/keeper/allocation.go#L14) as follows:
+
+```
+func (k Keeper) AllocateTokensFromPool(ctx sdk.Context, poolAddress string, totalPreviousPower int64, bondedVotes []abci.VoteInfo) {
+	rewardsCollectedInt := k.bankKeeper.GetAllBalances(ctx, poolAddress)
+	// rest in the same that `AllocateTokens` operates
+	...
+```
+
+In [TransferRewardsToFeeCollector](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/provider/keeper/distribution.go#L61) we then do:
+
+```
+for _, chain := range consumerChains {
+    ...
+    totalPower := 0
+    var []abci.VoteInfo votes
+    for _, val := range (validators that opted in or Top N) {
```

Comment:
How are we getting the validators in Top N? I was thinking that maybe it's just easier to store the consumer validator set for every consumer chain. When sending VSCPackets, we just update these valsets. Yeah, it's more storage, but it shouldn't be that much (unless we have 1000s of consumer chains).

Reply:
As discussed above, we are now going to implement top-n by just having a process that opts validators in automatically if they are in the top-n, so nothing else really needs to know about top-n, including this code.